swayfmt/src/comments.rs

```diff
@@ -0,0 +1,66 @@
+use std::fmt::Write;
+use sway_ast::token::Comment;
+use sway_types::Spanned;
+
+use crate::{formatter::FormattedCode, Formatter, FormatterError};
+
+// Given a start, an end and a CommentMap, return references to all comments
+// contained within a given start and end of a span, in an exclusive range.
+pub fn get_comments_between(start: usize, end: usize, formatter: &mut Formatter) -> Vec<&Comment> {
+    let mut comments = vec![];
+    let iter = formatter.comment_map.clone().into_keys();
+
+    for bs in iter {
+        if bs.start > start && bs.end < end {
+            if let Some(comment) = formatter.comment_map.get(&bs) {
+                comments.push(comment)
+            }
+        }
+    }
+
+    comments
+}
```

Comment:
This function currently collects all comments into a `Vec` before returning, which requires dynamically allocating the `Vec`. To avoid this, we could rewrite this function to return an `Iterator`, e.g. `-> impl Iterator<Item = &'_ Comment>`.

However, seeing as `comment_map` already provides a `range` method, we can likely omit this function altogether in favour of using something like:

```rust
formatter.comment_map.range(start..end)
```

Also just a note: `get` is generally reserved for datastructures that can `get` a reference to some element given an index or key. Here, it's safe to omit it from the function name (i.e. `comments_between`) as it's still clear what is being returned.

Reply:
Comment from above:

> Updated this to `comments_between()` - creating an iterator and returning it instead as you suggested below.

Reply:
I wonder if it's worth making `CommentMap` an actual type, e.g.

```rust
struct CommentMap(BTreeMap<ByteSpan, Comment>);
```
Then we could implement `Deref` and `DerefMut` targeting the inner `BTreeMap` to provide all the original methods, while adding functions like `comments_between` as actual *methods* on the comment map :thinking: